### DISCUSSION TITLE
Deprecate stateful protocol

### DISCUSSION THREAD

**Discussion by luckysori** [comit-network/baru#68] [Ideas]
Before coming up with a new design, we must understand how the current actor-state based approach is problematic.
 
We could then, for example, model the protocol as a series of transactions which either party can build, sign, broadcast.

**Comment by DeliciousHair** [comment]
Proposed design using PSETs:

![new_loan_protocol](https://user-images.githubusercontent.com/6540809/129130962-b2006a5e-b9a0-47ac-ba93-d89ac21146db.png)

**Comment by DeliciousHair** [comment]
As for why the current approach is up for an overhaul, this basically comes down to the fact that the current model does not cleanly separate the players (Borrower, Lender) and their associated states between `waves` and `baru`. While this works for now, it's not a terribly flexible approach and is difficult to scale / repurpose.

The proposal above essentially allows `waves` to provide the players and thus nodes to a computational graph, while `baru` is then able to provide the edges that act on elements script.
